### PR TITLE
test implementation

### PR DESCRIPTION
**Pull request by itaven**

### CODE REVIEW

File: streamer/src/test/utils/postgres/postgresQueries.js [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    const missedHashes = await getQuery(
+        postgres,
+        `SELECT id,parent_hash, prv_id,Prv_hash ` +
+        `FROM ` +
+        `(SELECT id,parent_hash, ` +
+        `Lag(hash) over (ORDER BY id) Prv_hash, ` +
+        `Lag(id) over (ORDER BY id) prv_id ` +
+        `FROM dot_polka.blocks) AS InnerQuery ` +
+        `WHERE parent_hash<>Prv_hash ` +
+        `ORDER BY id`
```
Comment (Nivelium) [inline streamer/src/test/utils/postgres/postgresQueries.js:113]: We can just wrap in all without concatenation multilines into
`

`

File: streamer/src/test/utils/postgres/postgresQueries.js [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+async function getQuery(postgres, queryString) {
+    return await postgres
+        .query(queryString)
+        .catch((err) => {
+            console.log(err.stack)
+            throw new Error(`Error executing query`)
+        })
+        .then((res) => {
+            return res
+        })
```
Comment (Nivelium) [inline streamer/src/test/utils/postgres/postgresQueries.js:17]: Let's use async/await + try/catch, instead of mixing promise and async/await syntax

**Review by Nivelium — CHANGES_REQUESTED** [review]

File: streamer/src/test/e2e/tests.js [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        } catch (err) {
+            console.log(err.stack)
+            throw new Error(`Connection error`)
+        }
+    })
+
+    it('Check for "chain" and "chainType"', async () => {
+        const [currentChainRaw, currentChainTypeRaw] = await Promise.all([
+            await nodeApi.rpc.system.chain(),
+            await nodeApi.rpc.system.chainType()
```
Comment (manylov) [inline streamer/src/test/e2e/tests.js:24]:
```suggestion
           nodeApi.rpc.system.chain(),
           nodeApi.rpc.system.chainType()
```

File: streamer/src/test/utils/postgres/postgresQueries.js [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+async function getMissedValidators(postgres) {
+    const missedValidators = await getQuery(
+        postgres,
+        `SELECT era FROM( ` +
+        `SELECT ${DB_SCHEMA}.eras.era,count(${DB_SCHEMA}.validators.era) ` +
+        `FROM ${DB_SCHEMA}.validators ` +
+        `FULL OUTER JOIN ${DB_SCHEMA}.eras on ${DB_SCHEMA}.validators.era=${DB_SCHEMA}.eras.era ` +
+        `GROUP BY ${DB_SCHEMA}.validators.era,${DB_SCHEMA}.eras.era ` +
+        `HAVING ${DB_SCHEMA}.eras.validators_active<>count(${DB_SCHEMA}.validators.era) ` +
+        `)t`
```
Comment (manylov) [inline streamer/src/test/utils/postgres/postgresQueries.js:39]:
```suggestion
        `SELECT era FROM( 
          SELECT ${DB_SCHEMA}.eras.era,count(${DB_SCHEMA}.validators.era) 
          FROM ${DB_SCHEMA}.validators 
          FULL OUTER JOIN ${DB_SCHEMA}.eras on ${DB_SCHEMA}.validators.era=${DB_SCHEMA}.eras.era
          GROUP BY ${DB_SCHEMA}.validators.era,${DB_SCHEMA}.eras.era 
          HAVING ${DB_SCHEMA}.eras.validators_active<>count(${DB_SCHEMA}.validators.era) 
          )t`
```

File: streamer/src/test/utils/postgres/postgresQueries.js [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+async function getMissedNominators(postgres) {
+    const missedNominators = await getQuery(
+        postgres,
+        `SELECT era FROM( ` +
+        `SELECT ${DB_SCHEMA}.eras.era,count(distinct ${DB_SCHEMA}.nominators.account_id) ` +
+        `FROM ${DB_SCHEMA}.nominators ` +
+        `FULL OUTER JOIN ${DB_SCHEMA}.eras on ${DB_SCHEMA}.nominators.era=${DB_SCHEMA}.eras.era ` +
+        `GROUP BY ${DB_SCHEMA}.nominators.era,${DB_SCHEMA}.eras.era ` +
+        `HAVING ${DB_SCHEMA}.eras.nominators_active<>count(distinct ${DB_SCHEMA}.nominators.account_id) ` +
+        `)t`
```
Comment (manylov) [inline streamer/src/test/utils/postgres/postgresQueries.js:56]:
```suggestion
        `SELECT era FROM(
          SELECT ${DB_SCHEMA}.eras.era,count(distinct ${DB_SCHEMA}.nominators.account_id)
          FROM ${DB_SCHEMA}.nominators
          FULL OUTER JOIN ${DB_SCHEMA}.eras on ${DB_SCHEMA}.nominators.era=${DB_SCHEMA}.eras.era
          GROUP BY ${DB_SCHEMA}.nominators.era,${DB_SCHEMA}.eras.era
          HAVING ${DB_SCHEMA}.eras.nominators_active<>count(distinct ${DB_SCHEMA}.nominators.account_id)
        )t`
```

File: streamer/src/test/e2e/tests.js [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        equal(missedHashes.breakCount, 0,
+            `There is ${missedHashes.breakCount} hashes sequence breaks; ` +
+                    `Breaks info: ${JSON.stringify(missedHashes.info)}`
+        )
+    })
+
+    it('Check node top block  with streamer top block', async () => {
+        const topBlock = await postgresQueries.getTopBlock(postgres)
+        const nodeTopBlock = await nodeApi.rpc.chain.getHeader()
+        const topBlockDifference = nodeTopBlock.number.toString() - topBlock
```
Comment (manylov) [inline streamer/src/test/e2e/tests.js:65]:
```suggestion
        const topBlockDifference = nodeTopBlock.number.toNumber() - topBlock
```

File: streamer/src/test/utils/postgres/postgresQueries.js [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+async function getErasWithRewardPointsInconsistencies(postgres) {
+    const inconsistencies = await getQuery(
+        postgres,
+        'SELECT era FROM( ' +
+        'SELECT dot_polka.eras.era,count( dot_polka.nominators.era),dot_polka.eras.nominators_active ' +
+        'FROM dot_polka.nominators ' +
+        'FULL OUTER JOIN dot_polka.eras on dot_polka.nominators.era=dot_polka.eras.era ' +
+        'GROUP BY dot_polka.nominators.era,dot_polka.eras.era ' +
+        'HAVING dot_polka.eras.nominators_active<>count(dot_polka.nominators.era) or dot_polka.eras.nominators_active is null ' +
+        ')t'
```
Comment (manylov) [inline streamer/src/test/utils/postgres/postgresQueries.js:130]:
```suggestion
        `SELECT era FROM(
        SELECT ${DB_SCHEMA}.eras.era,count( ${DB_SCHEMA}.nominators.era),${DB_SCHEMA}.eras.nominators_active 
        FROM ${DB_SCHEMA}.nominators 
        FULL OUTER JOIN ${DB_SCHEMA}.eras on ${DB_SCHEMA}.nominators.era=${DB_SCHEMA}.eras.era 
        GROUP BY ${DB_SCHEMA}.nominators.era,${DB_SCHEMA}.eras.era
        HAVING ${DB_SCHEMA}.eras.nominators_active<>count(${DB_SCHEMA}.nominators.era) or ${DB_SCHEMA}.eras.nominators_active is null
        )t`
```

File: streamer/src/test/e2e/tests.js [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    it('Check eras for low total reward points', async () => {
+        const erasCountWithLowTotalRewardPoints = await postgresQueries.getErasWithLowTotalRewardPoints(postgres)
+        equal(erasCountWithLowTotalRewardPoints.erasCount, 0,
+            `There is ${erasCountWithLowTotalRewardPoints.eras} eras with low total rewards points`)
+    })
+
+    it('Check rewards points in eras and sum of reward points in validators by eras', async () => {
+        const rewardPointsInconsistenciesEras = await postgresQueries.getErasWithRewardPointsInconsistencies(postgres)
+        equal(JSON.stringify(rewardPointsInconsistenciesEras.erasCount), 0,
+            `There is ${JSON.stringify(rewardPointsInconsistenciesEras.erasCount)} ` +
```
Comment (manylov) [inline streamer/src/test/e2e/tests.js:91]: Why JSON.stringify used here? 
Looks like rewardPointsInconsistenciesEras.erasCount has Number type

**Review by manylov — CHANGES_REQUESTED** [review]
Added some changes proposals and questions

File: streamer/src/test/e2e/tests.js (hunk 8 above)
Comment (itaven) [reply streamer/src/test/e2e/tests.js:91]: Yes,rewardPointsInconsistenciesEras.erasCount has a Number type and JSON.stringify shouldn't be used here.Got it.

File: streamer/src/test/utils/postgres/postgresQueries.js [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+
+async function getQuery(postgres, queryString) {
+    let queryResult
+    try {
+        queryResult = await postgres.query(queryString)
+    } catch (error) {
+        console.log(error.stack)
+        throw new Error(`Error executing query`)
+    }
+    return queryResult
```
Comment (Nivelium) [inline streamer/src/test/utils/postgres/postgresQueries.js:16]: It's better to use it inside try block.

File: streamer/src/test/utils/postgres/postgresQueries.js [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+async function getMissedBlocksCount(postgres) {
+    const missedBlocksCount = await getQuery(postgres, `SELECT max(id)-count(*) as missedBlocks FROM ${DB_SCHEMA}.blocks`)
+    return missedBlocksCount.rows[0].missedblocks
```
Comment (Nivelium) [inline streamer/src/test/utils/postgres/postgresQueries.js:21]: rows[0] possibly undefined.

File: streamer/src/test/utils/postgres/postgresQueries.js [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+async function getTopBlock(postgres) {
+    const topBlock = await getQuery(postgres, `SELECT max(id) as topBlock FROM ${DB_SCHEMA}.blocks`)
+    return topBlock.rows[0].topblock
```
Comment (Nivelium) [inline streamer/src/test/utils/postgres/postgresQueries.js:26]: rows[0] possibly undefined.

File: streamer/src/test/utils/postgres/postgresQueries.js [hunk 12]
```diff
@@ -0,0 +1,147 @@
+const {environment} = require('../../../environment')
+const DB_SCHEMA = environment.DB_SCHEMA
+const totalStakeThreshold = 1000000
+const totalRewardThreshold = 1000000
+const totalRewardPointsThreshold = 20
+
+
+async function getQuery(postgres, queryString) {
```
Comment (Nivelium) [inline streamer/src/test/utils/postgres/postgresQueries.js:8]: This function seems to be totally unnecessary as it's just an additional level of abstraction that messes.

File: streamer/src/test/utils/postgres/postgresQueries.js [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+    )
+    return {
+        count: missedNominators.rowCount,
+        eras: missedNominators.rows
+    }
+}
+
+async function getErasWithLowTotalStack(postgres) {
+    const erasWithLowTotalStack = await getQuery(postgres, `SELECT era FROM ${DB_SCHEMA}.eras WHERE total_stake<${totalStakeThreshold}`)
+    if (erasWithLowTotalStack.rows.length === 1 && erasWithLowTotalStack.rows[0].era === 0) {
```
Comment (Nivelium) [inline streamer/src/test/utils/postgres/postgresQueries.js:65]: Why do we need this?

File: streamer/src/test/utils/postgres/postgresQueries.js [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+        }
+    }
+    return {
+        erasCount: erasWithLowTotalStack.rows.length,
+        eras: erasWithLowTotalStack.rows
+    }
+}
+
+async function getErasWithLowTotalReward(postgres) {
+    const erasWithLowTotalReward = await getQuery(postgres, `SELECT era FROM ${DB_SCHEMA}.eras WHERE total_reward<${totalRewardThreshold}`)
```
Comment (Nivelium) [inline streamer/src/test/utils/postgres/postgresQueries.js:77]: Why do we need this?

File: streamer/src/test/utils/substrateInterface.js [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+        const wsProvider = new WsProvider(environment.SUBSTRATE_URI)
+        connection = await ApiPromise.create({provider: wsProvider})
+    }
+
+    return connection
+}
+
+async function apiDisconnect() {
+    if (connection != null) {
+        return await connection.disconnect()
```
Comment (Nivelium) [inline streamer/src/test/utils/substrateInterface.js:18]: unnecessary await

**Review by Nivelium — CHANGES_REQUESTED** [review]

File: streamer/src/test/utils/postgres/postgresQueries.js (hunk 13 above)
Comment (itaven) [reply streamer/src/test/utils/postgres/postgresQueries.js:65]: It is  an expected assertion for a case when the first era can contain low total stack

File: streamer/src/test/utils/postgres/postgresQueries.js (hunk 12 above)
Comment (itaven) [reply streamer/src/test/utils/postgres/postgresQueries.js:8]: IMO, that's reducing the number of lines for each query in functions

File: streamer/src/test/utils/postgres/postgresQueries.js (hunk 10 above)
Comment (itaven) [reply streamer/src/test/utils/postgres/postgresQueries.js:21]: Object missedBlocksCount return -  rows: [ { missedblocks: '3923' } ] .As far as I can tell it`s not undefined

File: streamer/src/test/utils/postgres/postgresQueries.js (hunk 14 above)
Comment (itaven) [reply streamer/src/test/utils/postgres/postgresQueries.js:77]: I didn’t catch that.

File: streamer/src/test/utils/substrateInterface.js (hunk 15 above)
Comment (itaven) [reply streamer/src/test/utils/substrateInterface.js:18]: Got it.

File: streamer/src/test/utils/postgres/postgresQueries.js (hunk 11 above)
Comment (itaven) [reply streamer/src/test/utils/postgres/postgresQueries.js:26]: The same as for https://github.com/p2p-org/polkadot-profit-transformer/pull/25#discussion_r622055305

File: streamer/src/test/utils/postgres/postgresQueries.js (hunk 12 above)
Comment (Nivelium) [reply streamer/src/test/utils/postgres/postgresQueries.js:8]: First, decreasing or increasing number of strings in the code is not the purpose itself. 
Second, this function itself is just a proxy wrapper for the native .query method of pg. If you switch it inside the functions where you call it, nothing will change. But it's just an additional abstraction layer that decreases readability of the code bringing no value. The native pg.query does exactly the same with throwing it's own error that is more clear.

File: streamer/src/test/utils/postgres/postgresQueries.js (hunk 10 above)
Comment (Nivelium) [reply streamer/src/test/utils/postgres/postgresQueries.js:21]: Okay, it's correct for aggregation functions.

File: streamer/src/test/utils/postgres/postgresQueries.js (hunk 13 above)
Comment (Nivelium) [reply streamer/src/test/utils/postgres/postgresQueries.js:65]: Why do you check if the lenght is equal to one? What if there is 2 eras with total stake below treshhold?

File: streamer/src/test/utils/postgres/postgresQueries.js (hunk 14 above)
Comment (Nivelium) [reply streamer/src/test/utils/postgres/postgresQueries.js:77]: Why do you check if the lenght is equal to one? What if there is 2 eras with total reward below treshhold?

**Review by Nivelium — CHANGES_REQUESTED** [review]

**Review by Nivelium — APPROVED** [review]

**Pull request closed without merging**